microsoft/MixedRealityToolkit-Unity
# Feature: Scene Transition Service

Overview
---
Adds a SceneTransitionService extension to MRTK to address #3899.

This extension simplifies the business of fading out a scene, displaying a progress indicator, loading a scene, then fading back in.

Scene operations are driven by the SceneSystem service. But in theory any Task-based operation could be used to drive a transition.

Testing Goals
---
- Use each of the progress indicator prefabs on device to check for rendering / material / animation bugs.
- Try different fade colors and fade lengths - are there bad combinations?
- Use fades on multiple camera targets - main cameras, UX cameras, custom combinations of both.
- Observe transitions between scenes with large amounts of content for progress indicator smoothness and other problems.
- Report any problems with MRTK services during transitions.

Details
---
For fading out / in, the service uses a class implementing ICameraFader. By default the CameraFaderQuad class is used, which instantiates a quad in front of the camera. Other methods (post FX, OnGUI, etc) could be substituted.

A few progress indicator options are included by default. Any prefab with a script implementing IProgressIndicator can be substituted.
- Progress Bar
- Rotating Object
- Rotating Orbs

![LoadingBar](https://user-images.githubusercontent.com/9789716/56066278-260f6c80-5d2c-11e9-84f0-a682636ae7ab.PNG)
![RotatingObject](https://user-images.githubusercontent.com/9789716/56066282-2871c680-5d2c-11e9-8eb6-4c042d96636e.PNG)
![LoadingOrbs](https://user-images.githubusercontent.com/9789716/56066284-29a2f380-5d2c-11e9-8863-0636989509e0.PNG)

Independent Transitions
---
You can also make FadeOut / FadeIn / ShowProgressIndicator calls independently of scene transitions. However these calls will not do anything if a scene transition is in progress.

Extension Service Discussion
---
Scripts and assets have been placed in MixedRealityToolkit/Extensions/SceneTransitionService/
This would be a good time to discuss:
- Is this the pattern we want to follow?
- Do extensions require their own assembly definition?
- Where to extension inspectors live?

Side note: this service was created using the extension creation wizard #3748
